gems/smithy-schema/lib/smithy-schema/document/deserializer.rb

```diff
@@ -0,0 +1,182 @@
+
+          @type_registry[document.discriminator]
+        end
+
+        def shape(ref, value, target = nil) # rubocop:disable Metrics/CyclomaticComplexity
+          case ref.shape
+          when StructureShape then structure(ref, value, target)
+          when UnionShape then union(ref, value, target)
+          when ListShape then list(ref, value, target)
+          when MapShape then map(ref, value, target)
+          when TimestampShape then timestamp(value)
+          when DocumentShape then document(value)
+          when BlobShape then Base64.strict_decode64(value)
+          when FloatShape then float(value)
+          else value
+          end
+        end
+
+        def document(values)
+          return values unless values.is_a?(Hash) && values.key?('__type')
+
+          msg = 'invalid document - document discriminator not found in type registry'
+          raise ArgumentError, msg unless @type_registry.key?(values['__type'])
+
+          shape_ref = ShapeRef.new(shape: @type_registry[values['__type']])
+          shape(shape_ref, values)
+        end
+
+        def float(value)
+          case value
+          when 'Infinity' then ::Float::INFINITY
+          when '-Infinity' then -::Float::INFINITY
+          when 'NaN' then ::Float::NAN
+          when nil then nil
+          else value.to_f
+          end
+        end
+
+        def list(ref, values, target = nil)
+          target = [] if target.nil?
+          values.each do |value|
+            next if value.nil? && !sparse?(ref.shape)
+
+            target << (value.nil? ? nil : shape(ref.shape.member, value))
+          end
+          target
+        end
+
+        def map(ref, values, target = nil)
+          target = {} if target.nil?
+          values.each do |key, value|
+            next if value.nil? && !sparse?(ref.shape)
+
+            target[key] = value.nil? ? nil : shape(ref.shape.value, value)
+          end
+          target
+        end
+
+        def structure(ref, values, target = nil)
+          return Smithy::Schema::EmptyStructure.new if ref.shape == Prelude::Unit
+
+          target = ref.shape.type.new if target.nil?
+          ref.shape.members.each do |member_name, member_ref|
+            name = member_ref.member_name
+            next unless values.key?(name)
+
+            target[member_name] = shape(member_ref, values[name])
+          end
+          target
+        end
+
+        def timestamp(value)
+          case value
+          when nil then nil
+          when Numeric
+            Time.at(value).utc
+          when /^[\d.]+$/
+            Time.at(value.to_f).utc
+          else
+            begin
+              fractional_time = Time.parse(value).to_f
+              Time.at(fractional_time).utc
+            rescue ArgumentError
+              raise "unhandled timestamp format `#{value}'"
+            end
+          end
+        end
+
+        def union(ref, values, target = nil) # rubocop:disable Metrics/AbcSize
```

Comment:
This one is wrong I think - just found this out today. Check out what is in my PR.